### PR TITLE
Miami

### PR DESCRIPTION
**Pull request by theR4K**
rework of processVehicle and processPlayerVehicleEngine functions in AudioLogic.cpp

### CODE REVIEW

File: src/audio/AudioLogic.cpp [hunk 1]
```diff
@@ -695,97 +697,140 @@ const bool hornPatternsArray[8][44] = {
...
-	case VEHICLE_TYPE_CAR:
-		UpdateGasPedalAudio((CAutomobile *)veh);
-		if (params.m_nIndex == RCBANDIT) {
-			ProcessModelCarEngine(&params);
+	player = FindPlayerPed();
+	playerVeh = FindPlayerVehicle();
+	if (playerVeh == NULL && player != NULL) {
+		attachedTo = player->m_attachedTo;
+		if (attachedTo != NULL && attachedTo->GetType() == ENTITY_TYPE_VEHICLE)
+			playerVeh = (CVehicle*)attachedTo;
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:715]: we prefer to use `nil` instead of `NULL`

File: src/audio/AudioLogic.cpp [hunk 2]
```diff
@@ -695,97 +697,140 @@ const bool hornPatternsArray[8][44] = {
...
+		|| CGame::currArea == AREA_BLOOD
+		|| CGame::currArea == AREA_DIRT
+		|| CGame::currArea == AREA_EVERYWHERE
+		|| CGame::currArea == AREA_MALL
+		|| CGame::currArea == AREA_MAIN_MAP) {
+		m_sQueueSample.m_vecPos = veh->GetPosition();
+		params.m_bDistanceCalculated = false;
+		params.m_pVehicle = veh;
+		params.m_fDistance = GetDistanceSquared(m_sQueueSample.m_vecPos);
+		params.m_pTransmission = veh->pHandling != NULL ? &veh->pHandling->Transmission : NULL;
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:728]: +1

File: src/audio/AudioLogic.cpp [hunk 3]
```diff
@@ -695,97 +697,140 @@ const bool hornPatternsArray[8][44] = {
...
-				break;
+		switch (params.m_VehicleType) {
+		case VEHICLE_TYPE_CAR:
+			automobile = (CAutomobile*)veh;
+			UpdateGasPedalAudio(&automobile->m_fGasPedalAudio, automobile->m_fGasPedal);
+			if (veh->m_modelIndex == MI_RCBANDIT || veh->m_modelIndex == MI_RCBARON) {
+				ProcessModelCarEngine(&params);//recheck
+				ProcessEngineDamage(&params);
+			} else if (veh->m_modelIndex == MI_RCRAIDER || veh->m_modelIndex == MI_RCGOBLIN) {
+				//ProcessModelHeliVehicle(this, &params);
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:750]: why?

File: src/audio/AudioLogic.cpp [hunk 4]
```diff
@@ -695,97 +697,140 @@ const bool hornPatternsArray[8][44] = {
...
+			ProcessHelicopter(&params); //recheck
+			ProcessVehicleOneShots(&params);
+			break;
+		case VEHICLE_TYPE_PLANE:
+			switch (params.m_nIndex) {
+			case AIRTRAIN:
+				ProcessJumbo(&params);//recheck
+				break;
+			case DEADDODO:
+				ProcessCesna(&params);//recheck
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:804]: imho it'd be better for these comments to be in header

File: src/audio/AudioLogic.cpp [hunk 5]
```diff
@@ -1060,6 +1105,16 @@ cAudioManager::ProcessVehicleEngine(cVehicleParams *params)
 			SampleManager.StopChannel(m_nActiveSamples);
 			return;
 		}
+		if (!params->m_bDistanceCalculated) {
+			if (params->m_fDistance <= 0.0)
+				m_sQueueSample.m_fDistance = 0.0;
+			else
+				m_sQueueSample.m_fDistance = sqrt(params->m_fDistance);
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:1112]: float based const

Also we have wrappers for sqrt: `Sqrt`.

File: src/audio/AudioLogic.cpp [hunk 6]
```diff
@@ -1398,138 +1506,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+	if (params->m_VehicleType == VEHICLE_TYPE_BIKE && bike->bExtraSpeed)
+		freqModifier += 1400;
+	tmp = 0;
 	engineSoundType = aVehicleSettings[params->m_nIndex].m_nBank;
-	soundOffset = 3 * (engineSoundType - CAR_SAMPLEBANKS_OFFSET);
+	soundOffset = 3 * (engineSoundType - 4);
+	err = false;
+	switch (soundOffset) {
+	case 0:
+		tmp = 2526;
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:1517]: it sounds like sound length?

File: src/audio/AudioLogic.cpp [hunk 7]
```diff
@@ -1305,91 +1359,145 @@ cAudioManager::ProcessCesna(cVehicleParams *params)
...
 	static bool bAccelSampleStopped = true;
 
 	lostTraction = false;
+	PizzaFaggBool = params->m_pVehicle->m_modelIndex == MI_PIZZABOY || params->m_pVehicle->m_modelIndex == MI_FAGGIO;
 	processedAccelSampleStopped = false;
 	if (bPlayerJustEnteredCar) {
-		bAccelSampleStopped = true;
-		bPlayerJustEnteredCar = false;
+		bAccelSampleStopped = 1;
+		bPlayerJustEnteredCar = 0;
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:1412]: why?

File: src/audio/AudioLogic.cpp [hunk 8]
```diff
@@ -1398,138 +1506,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
 					SampleManager.StopChannel(m_nActiveSamples);
 					bAccelSampleStopped = true;
 				}
-				AddPlayerCarSample(vol, freq, (engineSoundType - CAR_SAMPLEBANKS_OFFSET + SFX_CAR_REV_1), SAMPLEBANK_MAIN, 2, true);
+				if (params->m_pVehicle->bIsDrowning)
+					vol /= 4;
+				AddPlayerCarSample(vol, freq, engineSoundType + SFX_CAR_REV_1 - 4, SAMPLEBANK_MAIN, 2, true);
 			} else {
-				TranslateEntity(&m_sQueueSample.m_vecPos, &pos);
+				cAudioManager::TranslateEntity(&m_sQueueSample.m_vecPos, &pos);
```
Comment (ShFil119) [inline src/audio/AudioLogic.cpp:1670]: why?

File: src/audio/AudioManager.h [hunk 9]
```diff
@@ -358,6 +360,7 @@ class cAudioManager
 	void ProcessVehicle(CVehicle *vehicle);
 	bool ProcessVehicleDoors(cVehicleParams *params);
 	void ProcessVehicleEngine(cVehicleParams *params);
+	void UpdateGasPedalAudio(float* gasPedalAudio, float vehGasPedal);
```
Comment (Sergeanur) [inline src/audio/AudioManager.h:363]: This function takes (CVehicle*, int)

File: src/audio/AudioLogic.cpp [hunk 10]
```diff
@@ -695,97 +697,140 @@ const bool hornPatternsArray[8][44] = {
      false, false, false, false, true, true, true,  true,  true, true, true, true, true, true,  true,  true,  true, false, false, false, false, false},
 };
 
-
-void
-cAudioManager::ProcessVehicle(CVehicle *veh)
+void cAudioManager::ProcessVehicle(CVehicle* veh)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:700]: What's with the format change?

File: src/audio/AudioLogic.cpp [hunk 11]
```diff
@@ -1181,15 +1233,14 @@ cAudioManager::ProcessVehicleEngine(cVehicleParams *params)
 }
 
 void
-cAudioManager::UpdateGasPedalAudio(CAutomobile *automobile)
+cAudioManager::UpdateGasPedalAudio(float* gasPedalAudio, float vehGasPedal)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1236]: This function takes (CVehicle*, int)

File: src/audio/AudioLogic.cpp [hunk 12]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
 			freqModifier = -freqModifier;
 	} else
 		freqModifier = 0;
-
+	if (params->m_VehicleType == VEHICLE_TYPE_BIKE && bike->bExtraSpeed)
+		freqModifier += 1400;
+	gearSoundLength = 0;
 	engineSoundType = aVehicleSettings[params->m_nIndex].m_nBank;
-	soundOffset = 3 * (engineSoundType - CAR_SAMPLEBANKS_OFFSET);
+	soundOffset = 3 * (engineSoundType - 4);
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1509]: CAR_SAMPLEBANKS_OFFSET should stay and be adjusted

File: src/audio/AudioLogic.cpp [hunk 13]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+	err = false;
+	switch (soundOffset) {
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1511]: This was probably optimized engineSoundType switch (thus should be enum)

File: src/audio/AudioLogic.cpp [hunk 14]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+		gearSoundLength -= 1000;
+		if (CTimer::GetTimeInMilliseconds() - gearSoundStartTime > gearSoundLength) {
+			channelUsed = false;
+			gearSoundStartTime = CTimer::GetTimeInMilliseconds();
+		}
+	}
+	relativeVelocityChange = 2.0f * params->m_fVelocityChange / params->m_pTransmission->fMaxVelocity;
+	accelerationMultipler = clamp(relativeVelocityChange, 0.0f, 1.0f);
+	gasPedalAudio = accelerationMultipler;
+	switch (engineSoundType) {
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1570]: Should use enum (probably banks)

File: src/audio/AudioLogic.cpp [hunk 15]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+					gasPedalAudio = *gasPedalAudioPtr;
+				}
 			}
 			if (gasPedalAudio > 0.05f) {
 				freq = (5000.f * (gasPedalAudio - 0.05f) * 20.f / 19) + 19000;
-				if (engineSoundType == SAMPLEBANK_CAR_MACKTRUCK)
+				vol = (25.0f * (gasPedalAudio - 0.05f) * 20.f / 19) + 40;
+				if (params->m_pVehicle->bIsDrowning)
+					vol /= 4;
+				if (engineSoundType == SAMPLEBANK_CAR_COBRA)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1618]: This is still MACKTRUCK, not COBRA. Enum should be adjusted instead

File: src/audio/AudioLogic.cpp [hunk 16]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
 					freq /= 2;
-				AddPlayerCarSample((25.f * (gasPedalAudio - 0.05f) * 20.f / 19) + 40, freq, (soundOffset + SFX_CAR_FINGER_OFF_ACCEL_1), engineSoundType, 63,
-				                   false);
+				AddPlayerCarSample(vol, freq, soundOffset + SFX_CAR_FINGER_OFF_ACCEL_1, engineSoundType, 63, false);
 			}
 		}
 		freq = (10000.f * gasPedalAudio) + 22050;
-		if (engineSoundType == SAMPLEBANK_CAR_MACKTRUCK)
+		vol = 110 - (40.0f * gasPedalAudio);
+		if (engineSoundType == SAMPLEBANK_CAR_COBRA)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1625]: same as above

File: src/audio/AudioLogic.cpp [hunk 17]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+						*gasPedalAudioPtr *= 0.6f;
 					}
 					freqModifier = 0;
-					baseFreq = (15000.f * automobile->m_fGasPedalAudio) + 14000;
-					vol = (25.0f * automobile->m_fGasPedalAudio) + 60;
-				} else {
-					baseFreq = (8000.f * accelerationMultipler) + 16000;
-					vol = (25.0f * accelerationMultipler) + 60;
-					automobile->m_fGasPedalAudio = accelerationMultipler;
+					if (engineSoundType != 15 && engineSoundType != 16)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1649]: Should use enum

File: src/audio/AudioLogic.cpp [hunk 18]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+						baseFreq = (25000.0f * *gasPedalAudioPtr) + 14000;
+					else
+						baseFreq = (15000.0f * *gasPedalAudioPtr) + 14000;
+					vol = (25.0f * *gasPedalAudioPtr) + 60;
 				}
 				freq = freqModifier + baseFreq;
-				if (engineSoundType == SAMPLEBANK_CAR_MACKTRUCK)
+				if (engineSoundType == SAMPLEBANK_CAR_COBRA)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1656]: MACKTRUCK

File: src/audio/AudioLogic.cpp [hunk 19]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
-						else {
-							nCruising = 1;
-							break; // while was used just for this fucking place
-						}
+				if (channelUsed) {
+					SampleManager.SetChannelEmittingVolume(m_nActiveSamples, 120);
+					SampleManager.SetChannel3DPosition(m_nActiveSamples, pos.x, pos.y, pos.z);
+					SampleManager.SetChannel3DDistances(m_nActiveSamples, 50.0f, 12.5f);
+					freq = (GearFreqAdj[CurrentPretendGear] + freqModifier + 22050) ;
+					if (engineSoundType == 8)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1678]: MACKTRUCK

File: src/audio/AudioLogic.cpp [hunk 20]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+					if (!SampleManager.InitialiseChannel(m_nActiveSamples, soundOffset + SFX_CAR_ACCEL_1, SAMPLEBANK_MAIN))
+						return;
+					SampleManager.SetChannelLoopCount(m_nActiveSamples, 1);
+					SampleManager.SetChannelLoopPoints(m_nActiveSamples, 0, -1);
 
+					SampleManager.SetChannelEmittingVolume(m_nActiveSamples, 120);
+					SampleManager.SetChannel3DPosition(m_nActiveSamples, pos.x, pos.y, pos.z);
+					SampleManager.SetChannel3DDistances(m_nActiveSamples, 50.0f, 12.5f);
+					freq = (GearFreqAdj[CurrentPretendGear] + freqModifier + 22050);
+					if (engineSoundType == 8)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1697]: MACKTRUCK

File: src/audio/AudioLogic.cpp [hunk 21]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
-					freq /= 2;
-				SampleManager.SetChannelFrequency(m_nActiveSamples, freq);
-				if (!channelUsed) {
-					SampleManager.SetChannelReverbFlag(m_nActiveSamples, m_bDynamicAcousticModelingStatus != false);
-					SampleManager.StartChannel(m_nActiveSamples);
+					SampleManager.SetChannelEmittingVolume(m_nActiveSamples, 120);
+					SampleManager.SetChannel3DPosition(m_nActiveSamples, pos.x, pos.y, pos.z);
+					SampleManager.SetChannel3DDistances(m_nActiveSamples, 50.0f, 12.5f);
+					freq = (GearFreqAdj[CurrentPretendGear] + freqModifier + 22050);
+					if (engineSoundType == 8)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1717]: MACKTRUCK

File: src/audio/AudioLogic.cpp [hunk 22]
```diff
@@ -1398,138 +1502,272 @@ cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams *params, CAutomobile *
...
+				&& !lostTraction && currentGear >= params->m_pTransmission->nNumberOfGears - 1) {
+				if (accelerateState >= 220 && params->m_fVelocityChange + 0.001f >= velocityChangeForAudio) {
 					if (nCruising < 800)
 						++nCruising;
 				} else if (nCruising > 3) {
 					--nCruising;
 				}
 				freq = 27 * nCruising + freqModifier + 22050;
-				if (engineSoundType == SAMPLEBANK_CAR_MACKTRUCK)
+				if (engineSoundType == 8)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:1759]: MACKTRUCK

**Review by Sergeanur — CHANGES_REQUESTED** [review]

File: src/audio/AudioLogic.cpp (hunk 11 above)
Comment (theR4K) [reply src/audio/AudioLogic.cpp]: Actually it's takes (CVehicle*, eVehicleType)

Comment (Sergeanur) [reply src/audio/AudioLogic.cpp]: enums are reflected in symbols, so it was just int

File: src/audio/AudioLogic.cpp [hunk 23]
```diff
@@ -290,7 +290,10 @@ void
 cAudioManager::CalculateDistance(bool &distCalculated, float dist)
 {
 	if (!distCalculated) {
-		m_sQueueSample.m_fDistance = Sqrt(dist);
+		if (dist <= 0.0)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp:293]: Possibly better to reverse this check, also missing `f`

File: src/audio/AudioLogic.cpp [hunk 24]
```diff
@@ -1080,6 +1074,237 @@ cAudioManager::ProcessWetRoadNoise(cVehicleParams *params)
 	return true;
 }
 
+void
+cAudioManager::ProcessVehicleEngine1(cVehicleParams* params)
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp]: ProcessVehicleEngine**1**???

File: src/audio/AudioLogic.cpp [hunk 25]
```diff
@@ -1461,7 +1687,7 @@ void cAudioManager::ProcessPlayersVehicleEngine(cVehicleParams* params, CVehicle
 		case '4':
 			if (params->m_VehicleType != VEHICLE_TYPE_BIKE) {
 				wheelInUseCounter = 0;
-				for (uint8 i = 0; i < ARRAY_SIZE(wheelState); i++) {
+				for (uint8 i = 0; i < 4; i++) {
```
Comment (Sergeanur) [inline src/audio/AudioLogic.cpp]: Why change that?

Comment (theR4K) [reply src/audio/AudioLogic.cpp]: bcs sizeof(tWheelState*) it's not same with sizeof(array)

File: src/audio/AudioLogic.cpp (hunk 24 above)
Comment (theR4K) [reply src/audio/AudioLogic.cpp]: this commit created when I complete function, but not check if statements on errors. so 1 appears bcs I want to compare old and new function/

File: src/audio/AudioLogic.cpp (hunk 25 above)
Comment (Sergeanur) [reply src/audio/AudioLogic.cpp]: Ah, it's a pointer now, I see

**Review by Sergeanur — APPROVED** [review]

**Pull request merged**